worker/queue.go

```diff
@@ -0,0 +1,307 @@
+// It returns 0 if the task was not found.
+func (t tasks) Get(id uint64) TaskMeta {
+	if id == 0 || id == math.MaxUint64 {
+		return 0
+	}
+	t.logMu.Lock()
+	defer t.logMu.Unlock()
+	return TaskMeta(t.log.Get(id))
+}
+
+// set stores metadata for a given task ID.
+func (t tasks) set(id uint64, meta TaskMeta) {
+	t.logMu.Lock()
+	defer t.logMu.Unlock()
+	t.log.Set(id, uint64(meta))
+}
+
+// cleanup deletes all expired tasks.
+// This is not guaranteed to execute, it only runs if it hasn't run in a set interval.
+func (t tasks) cleanup() {
+	select {
+	case <-t.shouldCleanup.C:
+	default:
+		return
+	}
+
+	const ttl = 7 * 24 * time.Hour // 1 week
+	minTs := time.Now().UTC().Add(-ttl).Unix()
+	minMeta := uint64(minTs) << 32
+
+	t.logMu.Lock()
+	defer t.logMu.Unlock()
+	t.log.DeleteBelow(minMeta)
+}
+
+// run loops forever, running queued tasks one at a time. Any returned errors are logged.
+func (t tasks) run() {
+	for {
+		// If the server is shutting down, return immediately. Else, fetch a task from the queue.
+		var task taskRequest
+		select {
+		case <-x.ServerCloser.HasBeenClosed():
+			return
+		case task = <-t.queue:
+		}
+
+		// Fetch the task from the log. If the task isn't found, this means it has expired (older
+		// than taskTtl.
+		meta := TaskMeta(t.Get(task.id))
+		if meta == 0 {
+			glog.Errorf("task %#x: is expired, skipping", task.id)
+			continue
+		}
+		// Only proceed if the task is still queued. It's possible that the task got canceled
+		// before we were able to run it.
+		if status := meta.Status(); status != TaskStatusQueued {
+			glog.Errorf("task %#x: status is set to %s, skipping", task.id, status)
+		}
+		// Change the task status to RUNNING.
+		t.set(task.id, newTaskMeta(meta.Kind(), TaskStatusRunning))
+
+		// Run the task.
+		var status TaskStatus
+		if err := task.run(); err != nil {
+			status = TaskStatusFailed
+			glog.Errorf("task %#x: %s: %v", task.id, status, err)
+		} else {
+			status = TaskStatusSuccess
+			glog.Infof("task %#x: %s", task.id, status)
+		}
+
+		// Change the task status to SUCCESS / FAILED.
+		t.set(task.id, newTaskMeta(meta.Kind(), status))
+	}
+}
+
+// QueueBackup enqueues a backup request in the task queue.
+func (t tasks) QueueBackup(req *pb.BackupRequest) (uint64, error) {
+	return t.queueTask(req)
+}
+
+// QueueExport enqueues an export request in the task queue.
+func (t tasks) QueueExport(req *pb.ExportRequest) (uint64, error) {
+	return t.queueTask(req)
+}
+
+// queueTask enqueues a request of any type.
+// Don't use this function directly.
+func (t tasks) queueTask(req interface{}) (uint64, error) {
+	task := taskRequest{req: req}
+	for attempt := 0; ; attempt++ {
+		task.id = t.newId()
+		// z.Tree cannot store 0 or math.MaxUint64. Check that taskId is unique.
+		if task.id != 0 && task.id != math.MaxUint64 && t.Get(task.id) == 0 {
+			break
+		}
+		// Unable to generate a unique random number.
+		if attempt >= 8 {
+			t.rng.Seed(time.Now().UnixNano())
```

Comment:
Why are we re-seeding it?

Reply:
If we attempt to create a `taskId` 8 times and all 8 keys are taken, the RNG is likely broken. Since we're already seeding it at the start, this should never happen, but if it does, we re-seed it just in case.

Reply:
So, instead of dropping that request, shouldn't we retry at least once after re-seeding?

Reply:
I omitted it because it would add extra complexity to a scenario that would never happen in practice. I can add it in if you still feel like it would be good to have here.